crates/matrix-sdk/src/widget/widget_settings.rs

```diff
@@ -0,0 +1,217 @@
+        raw_url.push_str(&format!("&lang={}", url_props::LANGUAGE));
+        raw_url.push_str(&format!("&theme={}", url_props::CLIENT_THEME));
+        raw_url.push_str(&format!("&baseUrl={}", url_props::BASE_URL));
+
+        // Custom element call url parameters:
+        raw_url
+            .push_str(&format!("&parentUrl={}", encode(&parent_url.unwrap_or(element_call_url))));
+        if app_prompt.unwrap_or(false) {
+            raw_url.push_str("&embed=")
+        }
+        if hide_header.unwrap_or(true) {
+            raw_url.push_str("&hideHeader=")
+        }
+        if preload.unwrap_or(false) {
+            raw_url.push_str("&preload=")
+        }
+        if let Some(analytics_id) = analytics_id {
+            raw_url.push_str(&format!("&analyticsID={}", encode(&analytics_id)));
+        }
+        if let Some(scale) = font_scale {
+            raw_url.push_str(&format!("&fontScale={}", &scale.to_string()));
+        }
+        if skip_lobby.unwrap_or(false) {
+            raw_url.push_str("&skipLobby=");
+        }
+        if confine_to_room.unwrap_or(true) {
+            raw_url.push_str("&confineToRoom=");
+        }
+        if let Some(fonts) = fonts {
+            raw_url.push_str(&format!("&fonts={}", fonts.join(",")));
+        }
+
+        // for EC we always want init on load to be false
+        Self { id: widget_id, init_after_content_load: true, raw_url }
+    }
+
+    /// Create a new WidgetSettings instance
+    pub fn new(id: String, init_after_content_load: bool, raw_url: String) -> Self {
+        Self { id, init_after_content_load, raw_url }
+    }
+    // TODO: add From<WidgetStateEvent> so that WidgetSetting can be build
+    // by using the room state directly:
+    // Something like: room.get_widgets() -> Vec<WidgetStateEvent>
+}
+
+/// The set of settings and properties for the widget based on the client
+/// configuration. Those values are used generate the widget url.
+#[derive(Debug)]
+pub struct ClientProperties {
+    /// The client_id provides the widget with the option to behave differently
+    /// for different clients. e.g org.example.ios.
+    pub client_id: String,
+    /// The language the client is set to e.g. en-us.
+    pub language: LanguageTag,
+    /// A string describing the theme (dark, light) or org.example.dark.
+    pub theme: String,
+}
+impl ClientProperties {
+    /// Create client Properties with a String as the language_tag.
+    /// If a malformatted language_tag is provided it will default to en-US.
+    /// # Arguments
+    /// * `client_id` the client identifier. This allows widgets to adapt to
+    ///   specific clients (e.g. `io.element.web`)
+    /// * `language` the language that is used in the client. (default: `en-US`)
+    /// * `theme` the theme (dark, light) or org.example.dark. (default:
+    ///   `light`)
+    pub fn new(client_id: &str, language: Option<String>, theme: Option<String>) -> Self {
```

Comment:
I did not change this afterall. Reason beeing is that this would require the bindings create to also import LanguageTag. And for the ffi we ask for Strings anyways right?

Reply:
It's okay for the FFI API to use less strict types, but that doesn't mean we should use less strict types in the main crates.